### PR TITLE
SOLR-15961: Only consider the number in the SolrAuth header a timestamp if it's between a valid range of digits

### PR DESCRIPTION
**Pull request by tflobbe**

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/security/PKIAuthenticationPlugin.java [hunk 1]
```diff
@@ -75,6 +76,15 @@ public static void withServerIdentity(final boolean enabled) {
   }
 
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  /**
+   * If a number has less than this number of digits, it'll not be considered a timestamp.
+   */
+  private static final int MIN_TIMESTAMP_DIGITS = 10; // a timestamp of 9999999999 is year 1970
```
Comment (anshumg) [inline solr/core/src/java/org/apache/solr/security/PKIAuthenticationPlugin.java:83]: This range looks reasonable to parse a timestamp from now and years to come.

Considering it's expected to be a timestamp, we could have this be between 13 and 14 digits but guess that's just optimization that'd not really help with anything. Let's leave it at this :)

**Review by anshumg — APPROVED** [review]
LGTM

File: solr/core/src/test/org/apache/solr/security/TestPKIAuthenticationPlugin.java [hunk 2]
```diff
@@ -155,6 +161,70 @@ PublicKey getRemotePublicKey(String ignored) {
     mock1.close();
   }
 
+  @SuppressForbidden(reason="PKIAuthentication uses timestamps")
+  public void testParseCipher() {
+    for (String validUser: new String[]{"user1", "$", "some user","some 123"}) {
+      for (long validTimestamp: new long[]{System.currentTimeMillis(), 99999999999L, 9999999999999L}) {
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/security/TestPKIAuthenticationPlugin.java:167]: Instead of currentTimeMillis, could we do Instant.now.toEpochMill()?

File: solr/core/src/test/org/apache/solr/security/TestPKIAuthenticationPlugin.java [hunk 3]
```diff
@@ -155,6 +161,70 @@ PublicKey getRemotePublicKey(String ignored) {
...
+    assertNull(PKIAuthenticationPlugin.parseCipher(base64Cipher, aKeyPair.getPublicKey()));
+  }
+
+  @SuppressForbidden(reason="PKIAuthentication uses timestamps")
+  public void testParseCipherInvalidKey() {
+    String s = "user1 " + System.currentTimeMillis();
+    byte[] payload = s.getBytes(UTF_8);
+    byte[] payloadCipher = aKeyPair.encrypt(ByteBuffer.wrap(payload));
+    String base64Cipher = Base64.getEncoder().encodeToString(payloadCipher);
+    assertNull(PKIAuthenticationPlugin.parseCipher(base64Cipher, new CryptoKeys.RSAKeyPair().getPublicKey()));
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/security/TestPKIAuthenticationPlugin.java:206]: Is this the case that would occasionally get false positives? Can we hardcode the invalid key and timestamp combination here that leads to that?

File: solr/core/src/java/org/apache/solr/security/PKIAuthenticationPlugin.java (hunk 1 above)
Comment (tflobbe) [reply solr/core/src/java/org/apache/solr/security/PKIAuthenticationPlugin.java:83]: Yes, I thought about this too. At the end, I decided to keep the limits pretty open to handle only the bad parsing edge case, and let the regular expiration code handle the case where there is in fact a number. I'm happy to change the limits if you think it makes sense.

File: solr/core/src/test/org/apache/solr/security/TestPKIAuthenticationPlugin.java (hunk 3 above)
Comment (tflobbe) [reply solr/core/src/test/org/apache/solr/security/TestPKIAuthenticationPlugin.java:206]: I'll add a new test case with an example case of this particular issue

File: solr/core/src/java/org/apache/solr/security/PKIAuthenticationPlugin.java [hunk 4]
```diff
@@ -187,7 +198,7 @@ private static PKIHeaderData parseCipher(String cipher, PublicKey key) {
     }
     String s = new String(bytes, UTF_8).trim();
     int splitPoint = s.lastIndexOf(' ');
-    if (splitPoint == -1) {
+    if (splitPoint == -1 || s.length() - 1 - splitPoint < MIN_TIMESTAMP_DIGITS || s.length() - 1 - splitPoint > MAX_TIMESTAMP_DIGITS) {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/security/PKIAuthenticationPlugin.java:201]: For clarity let's store length-1-splitPoint as a variable?

**Review by madrob — APPROVED** [review]
LGTM, one minor idea, but fine to commit without it
